apps/ejabberd/src/ejabberd_hooks.erl

```diff
@@ -228,23 +228,26 @@ code_change(_OldVsn, State, _Extra) ->
 %%% Internal functions
 %%%----------------------------------------------------------------------
 
-run1([], _Hook, _Args) ->
-    ok;
-run1([{_Seq, Module, Function} | Ls], Hook, Args) ->
-    Res = if is_function(Function) ->
-                  safely:apply(Function, Args);
-             true ->
-                  safely:apply(Module, Function, Args)
+run1([], _Hook, Acc, _Args) ->
+    Acc;
+run1([{_Seq, Module, Function} | Ls], Hook, Acc, Args) ->
+    Res = case Function of
```

Comment:
How about `case` like the following:
```erlang
case is_function(Function) of
   true ->
       safely:apply(Function, [Acc | Args]);
  _ ->
       safely:apply(Module, Function, [Acc | Args])
end.
```

Looks simpler and easier to reason about.